cirq-core/cirq/ops/boolean_hamiltonian_test.py

```diff
@@ -98,3 +101,95 @@ def test_with_custom_names():
 
     with pytest.raises(ValueError, match='Length of replacement qubits must be the same'):
         original_op.with_qubits(q2)
+
+
+@pytest.mark.parametrize(
+    'n_bits,expected_hs',
+    [
+        (1, [(), (0,)]),
+        (2, [(), (0,), (0, 1), (1,)]),
+        (3, [(), (0,), (0, 1), (1,), (1, 2), (0, 1, 2), (0, 2), (2,)]),
+    ],
+)
+def test_gray_code_sorting(n_bits, expected_hs):
+    hs = []
+    for x in range(2 ** n_bits):
+        h = []
+        for i in range(n_bits):
+            if x % 2 == 1:
+                h.append(i)
+                x -= 1
+            x //= 2
+        hs.append(tuple(sorted(h)))
+    random.shuffle(hs)
+
+    sorted_hs = sorted(list(hs), key=functools.cmp_to_key(bh._gray_code_comparator))
+
+    np.testing.assert_array_equal(sorted_hs, expected_hs)
+
+
+@pytest.mark.parametrize(
+    'seq_a,seq_b,expected',
+    [
+        ((), (), 0),
+        ((), (0,), -1),
+        ((0,), (), 1),
+        ((0,), (0,), 0),
+    ],
+)
+def test_gray_code_comparison(seq_a, seq_b, expected):
+    assert bh._gray_code_comparator(seq_a, seq_b) == expected
+
+
+@pytest.mark.parametrize(
+    'input_cnots,input_flip_control_and_target,expected_simplified,expected_output_cnots',
+    [
+        # Empty inputs don't get simplified.
+        ([], False, False, []),
+        ([], True, False, []),
+        # Single CNOTs don't get simplified.
+        ([(0, 1)], False, False, [(0, 1)]),
+        ([(0, 1)], True, False, [(0, 1)]),
+        # Simplify away two CNOTs that are identical:
+        ([(0, 1), (0, 1)], False, True, []),
+        ([(0, 1), (0, 1)], True, True, []),
+        # Also simplify away if there's another CNOT in between.
+        ([(0, 1), (2, 1), (0, 1)], False, True, [(2, 1)]),
+        ([(0, 1), (0, 2), (0, 1)], True, True, [(0, 2)]),
+        # However, the in-between has to share the same target/control.
+        ([(0, 1), (0, 2), (0, 1)], False, False, [(0, 1), (0, 2), (0, 1)]),
+        ([(0, 1), (2, 1), (0, 1)], True, False, [(0, 1), (2, 1), (0, 1)]),
+    ],
+)
+def test_simplify_commuting_cnots(
+    input_cnots, input_flip_control_and_target, expected_simplified, expected_output_cnots
+):
+    actual_simplified, actual_output_cnots = bh._simplify_commuting_cnots(
+        input_cnots, input_flip_control_and_target
+    )
+    assert actual_simplified == expected_simplified
+    assert actual_output_cnots == expected_output_cnots
+
+
+@pytest.mark.parametrize(
+    'input_cnots,input_flip_control_and_target,expected_simplified,expected_output_cnots',
+    [
+        # Empty inputs don't get simplified.
+        ([], False, False, []),
+        ([], True, False, []),
+        # Single CNOTs don't get simplified.
+        ([(0, 1)], False, False, [(0, 1)]),
+        ([(0, 1)], True, False, [(0, 1)]),
+        # Simplify according to equation 11 of [4].
+        ([(2, 1), (2, 0), (1, 0)], False, True, [(1, 0), (2, 1)]),
+        ([(1, 2), (0, 2), (0, 1)], True, True, [(0, 1), (1, 2)]),
```

Comment:
This should include examples which have intervening, non-simplifying CNOTs, similar to the final four cases in the test above.

Reply:
Like this?

Reply:
I was also thinking of cases like:
```
# swap (2, 1) and (1, 0) around (2, 0)
([(2, 1), (2, 3), (2, 0), (1, 4), (1, 0)], False, True, [(1, 0), (2, 3), (1, 4), (2, 1)])
```
If I'm reading the code correctly this should work, but it's currently untested.

Reply:
Is the unit test what you had intended? The code actually doesn't simplify it, so I decided to dig a bit deeper. I built the two circuits:
`
          ┌──┐
0: ──────────────X───X───
                 │   │
1: ───X─────@────┼───@───
      │     │    │
2: ───@────@┼────@───────
           ││
3: ────────X┼────────────
            │
4: ─────────X────────────
          └──┘

0: ───X───────────
      │
1: ───@───@───X───
          │   │
2: ───@───┼───@───
      │   │
3: ───X───┼───────
          │
4: ───────X───────
`

Then, I computed their unitaries, and they seem to differ:
`U_1[4, 30]=(1+0j)` vs `U_2[4, 30]=0j`

Reply:
I'm sorry, my example was misconstructed. It should have been:

```
# swap (2, 1) and (1, 0) around (2, 0)
([(2, 1), (2, 3), (2, 0), (3, 0), (1, 0)], True, True, [(1, 0), (2, 3), (3, 0), (2, 1)])
```

Regardless of the example, the case I'm concerned about is one where there are intervening CNOTs between the pivot and one of the swapped CNOTs, but all of the intervening CNOTs share a target (or control) with both the pivot and the swapped CNOT on that side.